node/src/main.rs

```diff
@@ -598,133 +632,358 @@ async fn main() -> Result<(), Box<dyn Error>> {
+                        peer,
+                        message,
+                        connection_id,
+                    },
+                )) => {
+                    info!(
+                        peer = %peer,
+                        message = ?message,
+                        connection = ?connection_id,
+                        "Bead sync message received"
+                    );
+                    match message {
+                        request_response::Message::Request {
+                            request,
+                            request_id: _,
+                            channel,
+                        } => {
+                            // Handle the bead sync request here
+                            match request {
+                                bead::BeadRequest::GetBeads(hashes) => {
+                                    let mut beads = Vec::new();
+                                    {
+                                        let braid_lock = braid.read().await;
+                                        for hash in hashes.iter() {
+                                            if let Some(index) =
+                                                braid_lock.bead_index_mapping.get(hash)
+                                            {
+                                                if let Some(bead) = braid_lock.beads.get(*index) {
+                                                    beads.push(bead.clone());
+                                                }
+                                            }
+                                        }
+                                    }
+                                    //Sending all the beads requested in the hashes supplied during `GetData` request
+                                    swarm.behaviour_mut().respond_with_beads(channel, beads);
+                                }
+                                bead::BeadRequest::GetTips => {
+                                    let tips;
+                                    {
+                                        let braid_lock = braid.read().await;
+                                        tips = braid_lock
+                                            .tips
+                                            .iter()
+                                            .filter_map(|index| braid_lock.beads.get(*index))
+                                            .cloned()
+                                            .map(|bead| bead.block_header.block_hash())
+                                            .collect();
+                                    }
+                                    swarm.behaviour_mut().respond_with_tips(channel, tips);
+                                }
+                                bead::BeadRequest::GetGenesis => {
+                                    let genesis;
+                                    {
+                                        let braid_lock = braid.read().await;
```

Comment:
Don't necessarily need to lock the braid here. The genesis changes much more rarely. Maybe think of another way to do this. I think if we report accidentally a genesis from the previous epoch for $a$~1s during the (synchronous) switchover to a new epoch once every week or day, this is acceptable. The client ends up downloading the last day or week's beads but also will probably requesting the same info from other nodes, so this is extremely unlikely.

We can also insert heuristics to know whether we're near an epoch change, so the client knows and can react intelligently.